Add HttpKeepAliveHandler

otoolec opened the pull request:
Motivation:

As discussed in #5738, developers need to concern themselves with setting
connection: keep-alive on the response as well as whether to close a
connection or not after writing a response.  This leads to special keep-alive
handling logic in many different places.  The purpose of the HttpKeepAliveHandler
is to allow developers to add this handler to their pipeline and therefore
free themselves of having to worry about how Keep-Alive works.

Modifications:

Added HttpKeepAliveHandler to the io.netty.handler.codec.http package.

Result:

Developers can start using HttpKeepAliveHandler in their pipeline instead
of worrying about when to close a connection or set a Connection header
in their handler code.

willblackie commented on codec-http/src/main/java/io/netty/handler/codec/http/HttpKeepAliveHandler.java:
```diff
@@ -0,0 +1,132 @@
+ *  ...
+ *  p.addLast("serverCodec", new {@link HttpServerCodec}());
+ *  p.addLast("httpKeepAlive", <b>new {@link HttpKeepAliveHandler}()</b>);
+ *  p.addLast("serverCodec", new {@link HttpObjectAggregator}());
```
give this another name so the example code would work.. "aggregator"?

normanmaurer commented on codec-http/src/main/java/io/netty/handler/codec/http/HttpKeepAliveHandler.java:
```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright 2014 The Netty Project
```
2016

normanmaurer commented on codec-http/src/main/java/io/netty/handler/codec/http/HttpKeepAliveHandler.java:
```diff
@@ -0,0 +1,132 @@
+@ChannelHandler.Sharable
+public class HttpKeepAliveHandler extends ChannelDuplexHandler {
+
+    public static final String MULTIPART_PREFIX = "multipart";
```
private ?

normanmaurer commented on codec-http/src/main/java/io/netty/handler/codec/http/HttpKeepAliveHandler.java:
```diff
@@ -0,0 +1,132 @@
+    public static final String MULTIPART_PREFIX = "multipart";
+
+    /**
+     * Will close the channel if the connection couldn't be persisted.  Only works if HttpKeepAliveHandler is part of
```
Remove extra space before `Only`

normanmaurer commented on codec-http/src/main/java/io/netty/handler/codec/http/HttpKeepAliveHandler.java:
```diff
@@ -0,0 +1,132 @@
+        @Override
+        public void operationComplete(ChannelFuture channelFuture) throws Exception {
+            Boolean keepAlive = channelFuture.channel().attr(KEEP_ALIVE_ATTR).get();
+            if (!TRUE.equals(keepAlive) && channelFuture.channel().isOpen()) {
```
isActive() ?

normanmaurer commented on codec-http/src/main/java/io/netty/handler/codec/http/HttpKeepAliveHandler.java:
```diff
@@ -0,0 +1,132 @@
+     *
+     * @return true if the response has a self defined message length.
+     */
+    private boolean isSelfDefinedMessageLength(HttpResponse response) {
```
static ?

normanmaurer commented on codec-http/src/main/java/io/netty/handler/codec/http/HttpKeepAliveHandler.java:
```diff
@@ -0,0 +1,132 @@
+        super.write(ctx, msg, promise);
+    }
+
+    private boolean isKeepAliveEligible(HttpResponse response, Attribute<Boolean> keepAliveRequested) {
```
static ?

normanmaurer commented on codec-http/src/main/java/io/netty/handler/codec/http/HttpKeepAliveHandler.java:
```diff
@@ -0,0 +1,132 @@
+        return isContentLengthSet(response) || isTransferEncodingChunked(response) || isMultipart(response);
+    }
+
+    private boolean isMultipart(HttpResponse response) {
```
static ?

normanmaurer commented on codec-http/src/test/java/io/netty/handler/codec/http/HttpKeepAliveHandlerTest.java:
```diff
@@ -0,0 +1,99 @@
+package io.netty.handler.codec.http;
```
add license header

normanmaurer commented on codec-http/src/test/java/io/netty/handler/codec/http/HttpKeepAliveHandlerTest.java:
```diff
@@ -0,0 +1,99 @@
+        }
+
+
+        channel.writeInbound(request);
```
assert return value

normanmaurer commented on codec-http/src/test/java/io/netty/handler/codec/http/HttpKeepAliveHandlerTest.java:
```diff
@@ -0,0 +1,99 @@
+
+        channel.writeInbound(request);
+        channel.writeAndFlush(response);
+        HttpResponse writtenResponse = channel.readOutbound();
```
release response

normanmaurer commented on codec-http/src/test/java/io/netty/handler/codec/http/HttpKeepAliveHandlerTest.java:
```diff
@@ -0,0 +1,99 @@
+        HttpResponse writtenResponse = channel.readOutbound();
+
+        assertEquals("channel.isOpen", isKeepAliveResponseExpected, channel.isOpen());
+        assertEquals("response keep-alive", isKeepAliveResponseExpected, HttpUtil.isKeepAlive(writtenResponse));
```
call `channel.finish()` and assert return value

normanmaurer commented on codec-http/src/test/java/io/netty/handler/codec/http/HttpKeepAliveHandlerTest.java:
```diff
@@ -0,0 +1,99 @@
+                { false, HttpVersion.HTTP_1_1, null, NOT_SELF_DEFINED_MSG_LENGTH, null }                        // 11
+        });
+    }
+
```
remove empty line

normanmaurer commented on codec-http/src/test/java/io/netty/handler/codec/http/HttpKeepAliveHandlerTest.java:
```diff
@@ -0,0 +1,99 @@
+        default:
+            throw new IllegalArgumentException("selfDefinedMessageLength: " + setSelfDefinedMessageLength);
+        }
+
```
remove empty line

normanmaurer commented on codec-http/src/test/java/io/netty/handler/codec/http/HttpKeepAliveHandlerTest.java:
```diff
@@ -0,0 +1,99 @@
+    @Before
+    public void setUp() {
+        channel = new EmbeddedChannel(new HttpKeepAliveHandler());
+
```
remove empty line

otoolec commented:
@normanmaurer requested changes have been performed.  Tests are now passing.  Thanks for the feedback.

netkins commented on codec-http/src/main/java/io/netty/handler/codec/http/HttpKeepAliveHandler.java:
```diff
@@ -0,0 +1,132 @@
+                if (isKeepAliveEligible(response, keepAliveRequested)) {
+                    HttpUtil.setKeepAlive(response, true);
+                } else {
+                    HttpUtil.setKeepAlive(response, false);
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Unnecessary use of fully qualified name 'HttpUtil.setKeepAlive' due to existing static import 'io.netty.handler.codec.http.HttpUtil.*' [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AUnnecessaryFullyQualifiedName)

netkins commented on codec-http/src/main/java/io/netty/handler/codec/http/HttpKeepAliveHandler.java:
```diff
@@ -0,0 +1,132 @@
+        // read message and track if it was keepAlive
+        if (msg instanceof HttpRequest) {
+            HttpRequest request = (HttpRequest) msg;
+            ctx.channel().attr(KEEP_ALIVE_ATTR).set(HttpUtil.isKeepAlive(request));
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Unnecessary use of fully qualified name 'HttpUtil.isKeepAlive' due to existing static import 'io.netty.handler.codec.http.HttpUtil.*' [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AUnnecessaryFullyQualifiedName)

netkins commented on codec-http/src/main/java/io/netty/handler/codec/http/HttpKeepAliveHandler.java:
```diff
@@ -0,0 +1,132 @@
+                keepAliveRequested.set(HttpUtil.isKeepAlive(response));
+            } else {
+                if (isKeepAliveEligible(response, keepAliveRequested)) {
+                    HttpUtil.setKeepAlive(response, true);
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Unnecessary use of fully qualified name 'HttpUtil.setKeepAlive' due to existing static import 'io.netty.handler.codec.http.HttpUtil.*' [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AUnnecessaryFullyQualifiedName)

netkins commented on codec-http/src/main/java/io/netty/handler/codec/http/HttpKeepAliveHandler.java:
```diff
@@ -0,0 +1,132 @@
+            // Only mess with connection header if not already set
+            Attribute<Boolean> keepAliveRequested = ctx.channel().attr(KEEP_ALIVE_ATTR);
+            if (response.headers().contains(HttpHeaderNames.CONNECTION)) {
+                keepAliveRequested.set(HttpUtil.isKeepAlive(response));
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Unnecessary use of fully qualified name 'HttpUtil.isKeepAlive' due to existing static import 'io.netty.handler.codec.http.HttpUtil.*' [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AUnnecessaryFullyQualifiedName)

normanmaurer commented on codec-http/src/main/java/io/netty/handler/codec/http/HttpKeepAliveHandler.java:
```diff
@@ -0,0 +1,133 @@
+     * Only works if HttpKeepAliveHandler is part of the pipeline.
+     */
+    public static final ChannelFutureListener CLOSE_UNLESS_KEEP_ALIVE = new ChannelFutureListener() {
+        @SuppressWarnings("unused")
```
@otoolec remove ? Not sure why this is here.

otoolec replied:
I finally got the inspections working and this was flagged as unused.  I took the SuppressWarnings off and reran the inspection, but it isn't getting flagged this time, which makes sense since I'm clearly using it.  My guess is there was some kind of issue with my project setup.  Removing it now.

normanmaurer commented on codec-http/src/main/java/io/netty/handler/codec/http/HttpKeepAliveHandler.java:
```diff
@@ -0,0 +1,133 @@
+     *
+     * Only works if HttpKeepAliveHandler is part of the pipeline.
+     */
+    public static final ChannelFutureListener CLOSE_UNLESS_KEEP_ALIVE = new ChannelFutureListener() {
```
Also make it private ?

otoolec replied:
My initial implementation relied of users to add this listener manually instead of auto-adding it on LastHttpContent.  Made the change to reduce coupling but forgot to mark this as private.

normanmaurer commented on codec-http/src/main/java/io/netty/handler/codec/http/HttpKeepAliveHandler.java:
```diff
@@ -0,0 +1,133 @@
+     * Keep-alive only works if the client can detect when the message has ended without relying on the connection being
+     * closed.
+     * <p>
+     * See https://tools.ietf.org/html/rfc7230#section-6.3 See https://tools.ietf.org/html/rfc7230#section-3.3.3
```
Use `<a href="..."`

Scottmitch commented on codec-http/src/main/java/io/netty/handler/codec/http/HttpKeepAliveHandler.java:
```diff
@@ -0,0 +1,135 @@
+        @Override
+        public void operationComplete(ChannelFuture channelFuture) throws Exception {
+            Boolean keepAlive = channelFuture.channel().attr(KEEP_ALIVE_ATTR).get();
+            if (!TRUE.equals(keepAlive) && channelFuture.channel().isActive()) {
```
Is there a reason why we need to explicitly check `isActive()`?

normanmaurer replied:
@otoolec actually this is a good point... Calling `close()` on a closed channel is not a problem. You can remove the check.

otoolec replied:
Thanks for the feedback, I'll make this change.

Scottmitch commented on codec-http/src/main/java/io/netty/handler/codec/http/HttpKeepAliveHandler.java:
```diff
@@ -0,0 +1,135 @@
+            }
+        }
+        if (msg instanceof LastHttpContent) {
+            promise.addListener(CLOSE_UNLESS_KEEP_ALIVE);
```
Is it possible the connection will be prematurely closed when pipelineing is being used? For example if this handler is used on the server, the client issues 2 pipelined requests with the second request including a `Connection: Close` header, and the server reads both requests before sending any responses. In this case won't we close the connection after the server writes the first response?

normanmaurer replied:
good point... @otoolec I think a good way to solve this is just remove the `@Sharable` annotation and not use attributes at all. This way you could keep track of outstanding responses.

otoolec replied:
Makes sense.  I'll refactor to make sure we can handle clients that pipeline requests.  I'll also make sure we ignore 1xx responses when tracking outstanding responses.

otoolec commented:
@Scottmitch @normanmaurer I've added support for client pipelining, and tests.  The class is no longer sharable, but it does improve encapsulation as the channel attribute map is no longer used.

Should the class be renamed to `ServerHttpKeepAliveHandler` since it is only meant for server channels?

normanmaurer commented:
@otoolec LGTM... I think `HttpServerKeepAliveHandler` would be good as we also call it `HttpServerCodec` .

otoolec commented:
@normanmaurer Renamed and tests are passing, but let me know if I should rebase.

Scottmitch commented on codec-http/src/main/java/io/netty/handler/codec/http/HttpServerKeepAliveHandler.java:
```diff
@@ -0,0 +1,132 @@
+
+    // Using null as a way to know if this is our first time setting keepAliveRequested to false
+    private Boolean keepAliveRequested;
+    // Track pending responses to support client pipelining: https://www.w3.org/Protocols/rfc2616/rfc2616-sec8.html
```
rfc 2616 has been obsoleted by https://tools.ietf.org/html/rfc7230#section-6.3.2

Scottmitch commented on codec-http/src/main/java/io/netty/handler/codec/http/HttpServerKeepAliveHandler.java:
```diff
@@ -0,0 +1,132 @@
+        // read message and track if it was keepAlive
+        if (msg instanceof HttpRequest) {
+            final HttpRequest request = (HttpRequest) msg;
+            pendingResponses += 1;
```
consider using the increment operator instead `++pendingResponses`

otoolec replied:
I try to avoid the increment/decrement operators.  I can make this change if needed for editorial/consistency reasons.

Scottmitch replied:
> I try to avoid the increment/decrement operators.

Can you explain why?

Scottmitch commented on codec-http/src/main/java/io/netty/handler/codec/http/HttpServerKeepAliveHandler.java:
```diff
@@ -0,0 +1,132 @@
+
+    private void trackResponse(HttpResponse response) {
+        if (response.status().codeClass() != HttpStatusClass.INFORMATIONAL) {
+            pendingResponses -= 1;
```
consider using the decrement operator instead `--pendingResponses`

Scottmitch commented on codec-http/src/main/java/io/netty/handler/codec/http/HttpServerKeepAliveHandler.java:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    private boolean shouldKeepAlive() {
+        if (pendingResponses == 0) {
```
consider simplifying / refactoring:

```java
return pendingResponses == 0 ? keepAliveRequested : false;
```

otoolec replied:
Looks like Netkins didn't like that.  Simplified even further to `return pendingResponses != 0 || persistentConnection;`, which might be harder to understand.  Method name should help with that, but let me know if the logic is harder to understand and I can revert back to using the if / else to make reading easier.

Scottmitch replied:
+1 for Netkins :) what you have is better!

otoolec replied:
Would this be even more clear? Happy to make the change if you think it becomes even easier to read/understand. 

```java
return pendingResponses > 0 || persistentConnection;
```

Scottmitch commented on codec-http/src/main/java/io/netty/handler/codec/http/HttpServerKeepAliveHandler.java:
```diff
@@ -0,0 +1,132 @@
+            }
+            // Server might think it can keep connection alive, but we should fix response header if we know better
+            if (!shouldKeepAlive()) {
+                setKeepAlive(response, false);
```
what if the user explicitly put a `Connection: close` header on the response, and does not plan on returning a response for any other pipelined requests? IIRC the RFC permits this and the server can just close the connection at its discretion.

otoolec replied:
This is already handled by setting `pendingResponses` to 0 above.  I'll add some tests to show what I mean.

Scottmitch commented on codec-http/src/main/java/io/netty/handler/codec/http/HttpServerKeepAliveHandler.java:
```diff
@@ -0,0 +1,132 @@
+
+    private boolean shouldKeepAlive() {
+        if (pendingResponses == 0) {
+            return keepAliveRequested;
```
`keepAliveRequested` is a `Boolean` and may be `null` ... possible NPE here?

Also can `keepAliveRequested` just be a `boolean`?

otoolec replied:
I'm not sure what I was trying to do here anymore.  Maybe it evolved from when it was an Attribute?  I know I wanted to use `null` to know if things had been initialized or not, but the code sure is confusing now.  

I'll rename the variable to persistentConnection and initialize it to `true` (`boolean`).  Thanks for making me look at this again, it was pretty ugly.

Scottmitch commented on codec-http/src/main/java/io/netty/handler/codec/http/HttpServerKeepAliveHandler.java:
```diff
@@ -0,0 +1,132 @@
+
+    private static boolean isMultipart(HttpResponse response) {
+        String contentType = response.headers().get(HttpHeaderNames.CONTENT_TYPE);
+        return contentType != null && contentType.startsWith(MULTIPART_PREFIX);
```
IIRC header values should be compared in a case-insensitive manner.

otoolec replied:
Yes, will fix. 
https://tools.ietf.org/html/rfc7231#section-3.1.1.1

Scottmitch commented on codec-http/src/main/java/io/netty/handler/codec/http/HttpServerKeepAliveHandler.java:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    private void trackResponse(HttpResponse response) {
+        if (response.status().codeClass() != HttpStatusClass.INFORMATIONAL) {
```
nit: `!isInformational(response)`

otoolec commented:
@Scottmitch I think I've addressed all of your comments.  Thanks for the thorough review!

Should I be force pushing these updates to keep things easy to cherry-pick or does it make more sense for me to push individual update commits during these code reviews?

netkins commented on codec-http/src/main/java/io/netty/handler/codec/http/HttpServerKeepAliveHandler.java:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    private boolean shouldKeepAlive() {
+        return pendingResponses == 0 ? persistentConnection : true;
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Ternary operators that can be simplified with || or && [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3ASimplifiedTernary)

Scottmitch commented on codec-http/src/main/java/io/netty/handler/codec/http/HttpServerKeepAliveHandler.java:
```diff
@@ -0,0 +1,127 @@
+        if (msg instanceof HttpRequest) {
+            final HttpRequest request = (HttpRequest) msg;
+            if (persistentConnection) {
+                pendingResponses += 1;
```
The other comment https://github.com/netty/netty/pull/5788#discussion_r78794055 was collapsed but I still have a question:

@otoolec:
> I try to avoid the increment/decrement operators. I can make this change if needed for editorial/consistency reasons.

me:

> Can you explain why?

Scottmitch replied:
also we conditionally increment `pendingResponses` based upon `persistentConnection` but do not consider `persistentConnection` when decremented ... can you make this consistent?

otoolec replied:
I listened to a talk by Douglas Crockford that left me convinced that using them wasn't worth it.  I think it boiled down to:
- Can lead to bugs that are tricky for some developers to understand
- Code is written once, read many times by many people, so keep it simple for them
- Easier to update if you want to increment by more than one
- Better to get into the habit of not using it instead of deciding where it is safe to do so.

otoolec replied:
I used to unconditionally increment `pendingResponses` but it made test case 7 fail once I started using it in the pipeline test.  If I make them consistent it will cause either test 7 to fail or 0-7 (except 2 and 4).  

The idea is once we know that this is no longer a persistent connection (for example the client sent `Connection: close`) then we should stop counting requests that are coming in.  It shouldn't really happen, but maybe a client sends a request to close the connection and then accidentally sends another request, in that case we will close the connection anyway as requested.

Scottmitch replied:
can we just unconditionally increment the count then? Seems unexpected to have `pendingResponses` potentially go negative

Scottmitch replied:
sgtm

Scottmitch commented on codec-http/src/main/java/io/netty/handler/codec/http/HttpServerKeepAliveHandler.java:
```diff
@@ -0,0 +1,127 @@
+            final HttpResponse response = (HttpResponse) msg;
+            trackResponse(response);
+            // Assume the response writer knows if they can persist or not and sets isKeepAlive on the response
+            if (!isKeepAlive(response) || !isSelfDefinedMessageLength(response)) {
```
https://tools.ietf.org/html/rfc7230#section-6.3
>HTTP/1.1 defaults to the use of "persistent connections"

Does this statement account for that behavior?

Scottmitch replied:
all good ... HttpUtil.isKeepAlive(..) should handle this.

Scottmitch commented on codec-http/src/main/java/io/netty/handler/codec/http/HttpServerKeepAliveHandler.java:
```diff
@@ -0,0 +1,127 @@
+            }
+        }
+        if (msg instanceof LastHttpContent && !shouldKeepAlive()) {
+            promise.addListener(ChannelFutureListener.CLOSE);
```
should we also ensure the response has the `Connection: close` header?

otoolec replied:
That should be happening on line 80 above.   I don't think there is an ordering problem since `trackResponse` is called before the check on line 80.  I had to split things out like this because we may be getting **HttpResponse** written one or more events before the **LastHttpContent** is written.

Scottmitch replied:
thanks for clarifying ... lgtm

Scottmitch commented:
thanks @otoolec !

@normanmaurer - lgtm please merge when satisfied.

normanmaurer commented:
Cherry-picked into 4.1 (c57d4bed91eff61515f25713618101cc4072ade5) and 4.0 (62d75cc10a1fd90732973215e65370f1fd457736)

normanmaurer commented:
@otoolec thanks again!

otoolec commented:
Thanks for the great feedback @Scottmitch @normanmaurer !
